### PR TITLE
Blocks API: Add 'lock' attribute during block registration

### PR DESCRIPTION
**Pull request by Mamaduka**
## Description
Treat `lock` as a "build-in" attribute and include it during block registration. Removes filter previously used for adding this attribute.

Based on https://github.com/WordPress/gutenberg/pull/32457#discussion_r700917579.
Part of #29864.

## How has this been tested?
Check that unit tests are passing.

1. Use the snippet below to insert the locked block.
2. Block movers shouldn't be visible.
3. Remove block option isn't available.

## Example block
```html
<!-- wp:heading {"lock":{"remove":true,"move":true}} -->
<h2>Locked Heading</h2>
<!-- /wp:heading --></div>
```

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/blocks/src/api/registration.js [hunk 1]
```diff
@@ -365,6 +366,17 @@ export function registerBlockType( blockNameOrMetadata, settings ) {
 		return;
 	}
 
+	// Handle `lock` as a "build-in" attribute.
+	// But allow blocks to specify their own with defaults.
+	if ( ! has( settings.attributes, [ 'lock', 'type' ] ) ) {
```
Comment (Mamaduka) [inline packages/blocks/src/api/registration.js:371]: I noticed while updating unit tests that there's an expectation for empty attributes if the block doesn't have them.

Should I update the condition and only add the `lock` attribute if attributes aren't empty?

**Comment by gziolo** [comment]
Changes on the client are not enough if we want to make the `lock` attribute always present in the block definition. I'm not fully convinced that only the `lock` should have special handling. There is a long list of attributes we inject with hooks through `supports` that we should also revisit before committing to anything.

**Comment by Mamaduka** [comment]
Good point, @gziolo.

Currently, we inject at least nine attributes using filters. I also would prefer a unified way to handle "build-in" attribues.

**Comment by youknowriad** [comment]
The difference is that the attributes added by the "hooks" are not used in other parts of the code other than the "hooks" so even if they live in `block-editor` they can be really considered "hooks", `lock` on the other side is used in the framework itself.

(`align` is a bit between both because the current implementation is kind of broken for old themes)

**Comment by gziolo** [comment]
> lock on the other side is used in the framework itself.

I guess the framework could provide a fallback when an attribute is not set, too. I don't have a strong opinion on that though. I don't know why `lock` has to be so deeply integrated with the framework as it's purely edit UI based feature.

On the PHP side, we will have to change a few things starting from the default value for `attributes` in `WP_Block_Type` class:

https://github.com/WordPress/wordpress-develop/blob/ed9f437fc098ba2c9ba18eda0fa341ecd0476ca9/src/wp-includes/class-wp-block-type.php#L134-L140

If I recall correctly, there was a valid reason to use `null` rather than an empty array there. We will have to double check.

**Comment by senadir** [comment]
@gziolo and @youknowriad is this PR blocked? I see the logic of having this as a built in but why should it be added to the registration process? can't the logic that relies on it just assume if `lock` isn't there, it's false?

So I'm leaning more toward this:

> I guess the framework could provide a fallback when an attribute is not set, too. I don't have a strong opinion on that though. I don't know why lock has to be so deeply integrated with the framework as it's purely edit UI based feature.

**Comment by paaljoachim** [comment]
Is this a feature/enhancement that should be added to WordPress 6.0?

**Comment by Mamaduka** [comment]
I think we can close this one. The current method for registering the attribute works well.

**Comment by gziolo** [comment]
> I think we can close this one. The current method for registering the attribute works well.

We can always reopen if necessary 👍

**Pull request closed without merging**